http/src/lib.rs

```diff
@@ -581,12 +575,46 @@ impl Server {
 			executor.wait();
 		}
 	}
+
+	/// Returns a handle to the server that can be used to close it while another thread is
+    /// blocking in `wait`.
+    pub fn close_handle(&mut self) -> CloseHandle {
+		let executor_close: Option<Vec<_>> = self.executor.as_mut().map(|executors| {
+			executors.iter_mut().map(|executor| executor.close_handle()).collect()
+		});
+
+        CloseHandle {
+            executor_close,
+            close: self.close.clone(),
+        }
+    }
 }
 
 impl Drop for Server {
 	fn drop(&mut self) {
-		self.executor.take().map(|executors| {
-			for executor in executors { executor.close(); }
-		});
+		self.close_handle().close();
 	}
 }
+
+/// A handle that allows closing of a server even if it owned by a thread blocked in `wait`.
+pub struct CloseHandle {
```

Comment:
I think we should implement/derive `Clone` for the `CloseHandle`.

Reply:
futures::Sender, doesn't implement Clone(), so how should we approach this?

Reply:
Do you mean the `executor_close` can not be clonded?

Reply:
@carllin it's wrapped in `Arc<Mutex` anyway, so it doesn't really matter. If derive doesn't work then you need to implement `Clone` manually (and just clone the `Arc`)

Reply:
@bkchr, @tomusdrw, yes the executor_close cannot be cloned.

Reply:
@tomusdrw, in the current WS implementation, close() also closes the executor: https://github.com/paritytech/jsonrpc/blob/master/ws/src/server.rs#L149